src/node_options.h

```diff
@@ -0,0 +1,351 @@
+  bool no_deprecation = false;
+  bool no_force_async_hooks_checks = false;
+  bool no_warnings = false;
+  bool pending_deprecation = false;
+  bool preserve_symlinks = false;
+  bool preserve_symlinks_main = false;
+  bool prof_process = false;
+  std::string redirect_warnings;
+  bool throw_deprecation = false;
+  bool trace_deprecation = false;
+  bool trace_sync_io = false;
+  bool trace_warnings = false;
+  std::string userland_loader;
+
+  bool syntax_check_only = false;
+  bool has_eval_string = false;
+  std::string eval_string;
+  bool print_eval = false;
+  bool force_repl = false;
+
+  std::vector<std::string> preload_modules;
+
+  std::vector<std::string> user_argv;
+
+  inline DebugOptions* get_debug_options();
+};
+
+class PerIsolateOptions {
+ public:
+  std::shared_ptr<EnvironmentOptions> per_env { new EnvironmentOptions() };
+  bool track_heap_objects = false;
+
+  inline EnvironmentOptions* get_per_env_options();
+};
+
+class PerProcessOptions {
+ public:
+  std::shared_ptr<PerIsolateOptions> per_isolate { new PerIsolateOptions() };
+
+  std::string title;
+  std::string trace_event_categories;
+  std::string trace_event_file_pattern = "node_trace.${rotation}.log";
+  int64_t v8_thread_pool_size = 4;
+  bool zero_fill_all_buffers = false;
+
+  std::vector<std::string> security_reverts;
+  bool print_help = false;
+  bool print_v8_help = false;
+  bool print_version = false;
+
+#ifdef NODE_HAVE_I18N_SUPPORT
+  std::string icu_data_dir;
+#endif
+
+  // TODO(addaleax): Some of these could probably be per-Environment.
+#if HAVE_OPENSSL
+  std::string openssl_config;
+  std::string tls_cipher_list = DEFAULT_CIPHER_LIST_CORE;
+#ifdef NODE_OPENSSL_CERT_STORE
+  bool ssl_openssl_cert_store = true;
+#else
+  bool ssl_openssl_cert_store = false;
+#endif
+  bool use_openssl_ca = false;
+  bool use_bundled_ca = false;
+#if NODE_FIPS_MODE
+  bool enable_fips_crypto = false;
+  bool force_fips_crypto = false;
+#endif
+#endif
+
+  inline PerIsolateOptions* get_per_isolate_options();
+};
+
+// The actual options parser, as opposed to the structs containing them:
+
+HostPort SplitHostPort(const std::string& arg, std::string* error);
+
+enum OptionEnvvarSettings {
+  kAllowedInEnvironment,
+  kDisallowedInEnvironment
+};
+
+enum OptionType {
+  kNoOp,
+  kV8Option,
+  kBoolean,
+  kInteger,
+  kString,
+  kHostPort,
+  kStringList,
+};
+
+template <typename Options>
+class OptionsParser {
+ public:
+  virtual ~OptionsParser() {}
+
+  typedef Options TargetType;
+
+  struct NoOp {};
+  struct V8Option {};
+
+  // TODO(addaleax): A lot of the `std::string` usage here could be reduced
+  // to simple `const char*`s if it's reasonable to expect the values to be
+  // known at compile-time.
+
+  // These methods add a single option to the parser. Optionally, it can be
+  // specified whether the option should be allowed from environment variable
+  // sources (i.e. NODE_OPTIONS).
+  void AddOption(const std::string& name,
```

Comment:
Why can't these be templates?

Reply:
The way this is done is because we use it to infer the `info.type` value. We could do something with templates but I think it would mean we’d still use a list of template specializations?